src/java.base/share/classes/jdk/internal/foreign/SegmentBulkOperations.java

```diff
@@ -0,0 +1,271 @@
+            assert remaining < 8 : "remaining greater than 7: " + remaining;
+            i = bytes - remaining;
+            return mismatch(src, srcFromOffset + i, dst, dstFromOffset + i, i, (int) remaining, srcAndDstBytesDiffer);
+        }
+    }
+
+    @ForceInline
+    private static long mismatch(AbstractMemorySegmentImpl src, long srcFromOffset,
+                                 AbstractMemorySegmentImpl dst, long dstFromOffset,
+                                 long start, int bytes, boolean srcAndDstBytesDiffer) {
+        final int limit = (int) (bytes & (NATIVE_THRESHOLD_MISMATCH - 8));
+        int offset = 0;
+        for (; offset < limit; offset += 8) {
+            final long s = SCOPED_MEMORY_ACCESS.getLong(src.sessionImpl(), src.unsafeGetBase(), src.unsafeGetOffset() + srcFromOffset + offset);
+            final long d = SCOPED_MEMORY_ACCESS.getLong(dst.sessionImpl(), dst.unsafeGetBase(), dst.unsafeGetOffset() + dstFromOffset + offset);
+            if (s != d) {
+                return start + offset + mismatch(s, d);
+            }
+        }
+        int remaining = bytes - offset;
+        // 0...0X00
+        if (remaining >= 4) {
+            final int s = SCOPED_MEMORY_ACCESS.getInt(src.sessionImpl(), src.unsafeGetBase(), src.unsafeGetOffset() + srcFromOffset + offset);
+            final int d = SCOPED_MEMORY_ACCESS.getInt(dst.sessionImpl(), dst.unsafeGetBase(), dst.unsafeGetOffset() + dstFromOffset + offset);
+            if (s != d) {
+                return start + offset + mismatch(s, d);
+            }
+            offset += 4;
+            remaining -= 4;
+        }
+        // 0...00X0
+        if (remaining >= 2) {
+            final short s = SCOPED_MEMORY_ACCESS.getShort(src.sessionImpl(), src.unsafeGetBase(), src.unsafeGetOffset() + srcFromOffset + offset);
+            final short d = SCOPED_MEMORY_ACCESS.getShort(dst.sessionImpl(), dst.unsafeGetBase(), dst.unsafeGetOffset() + dstFromOffset + offset);
+            if (s != d) {
+                return start + offset + mismatch(s, d);
+            }
+            offset += 2;
+            remaining -= 2;
+        }
+        // 0...000X
+        if (remaining == 1) {
+            final byte s = SCOPED_MEMORY_ACCESS.getByte(src.sessionImpl(), src.unsafeGetBase(), src.unsafeGetOffset() + srcFromOffset + offset);
+            final byte d = SCOPED_MEMORY_ACCESS.getByte(dst.sessionImpl(), dst.unsafeGetBase(), dst.unsafeGetOffset() + dstFromOffset + offset);
+            if (s != d) {
+                return start + offset;
+            }
+        }
+        return srcAndDstBytesDiffer ? (start + bytes) : -1;
+        // We have now fully handled 0...0X...XXXX
+    }
+
+    @ForceInline
+    private static int mismatch(long first, long second) {
+        final long x = first ^ second;
+        return (Architecture.isLittleEndian()
+                ? Long.numberOfTrailingZeros(x)
+                : Long.numberOfLeadingZeros(x)) / 8;
```

Comment:
Would there be value in having a constant LongToIntFunction lambda to capture this logic?

```
private static final LongToIntFunction LEADING_ZEROS = Architecture.isLittleEndian() ? Long::numberOfTrailingZeros : Long::numberOfLeadingZeros;
```

Reply:
part of the value would be if it allowed just removing this method entirely and moving the logic above:

```
            if (s != d) {
                return start + offset + mismatch(s, d);
            }
```

would become:

```
            if (s != d) {
                return start + offset + (LONG_LEADING_ZEROS.apply(s^d) >>> 3);
            }
```